### ISSUE TITLE
',' is invalid after a single JSON value. Expected end of data.

### ISSUE THREAD

**Issue by ericwj** [dotnet/corefx#39843]
I am reading a single complex object from a `FileStream` using `JsonSerializer.ReadAsync<T>(stream, options, token)` and I seriously just want the one value, but I am getting this exception.

```
',' is invalid after a single JSON value. Expected end of data.
```

The JSON in the file is preceded by bytes which I have skipped or already processed and it is followed by more bytes which I can skip or process separately. However I get stuck by this exception and don't get the value that is presumably correctly read.

I have no idea how long the value is. Pre-parsing the JSON at a lower level to just get the byte length is a bad solution.

Similar to #38670.

**Comment by ericwj** [comment]
I have tried setting `JsonSerializerOptions.AllowTrailingCommas` but that didn't make a difference.

**Comment by ahsonkhan** [comment]
@ericwj can you share your input JSON payload and also the .NET object model you are deserializing to? 

Also, what version of dotnet sdk are you using (share the output of `dotnet --info`)? Can you try with preview 9 bits (or latest nightly)?

Please try the nightly SDK/runtime with the fix to verify it works (https://github.com/dotnet/core-sdk#installers-and-binaries), or wait till preview 9 ships. Alternatively, you could reference the latest S.T.Json NuGet package. For example:

```csproj
  <ItemGroup>
    <PackageReference Include="System.Text.Json" Version="4.6.0-preview9.19413.13" />
    <PackageReference Include="System.Text.Encodings.Web" Version="4.6.0-preview9.19413.13" />
  </ItemGroup>
```

**Comment by ericwj** [comment]
Yes I am trying to read in a streaming fashion, so there is JSON following other JSON. The files can become fairly big.

I think I might be able to use a `Pipe` and an `Utf8JsonReader` to do the job, although working with `SequencePosition` is awkward and `JsonSerializer.Deserialize<T>(ref reader, ...)` will throw exceptions as long as the buffer isn't valid yet while I search for the next '}'. Wouldn't it be better if there was a `bool TryDeserialize<T>` anyway?

**Comment by ahsonkhan** [comment]
> Wouldn't it be better if there was a `bool TryDeserialize<T>` anyway?

What would you expect TryDeserialize to do/return in your scenario where you have multiple json payloads within the file and it fails to deserialize? The deserialize call today throws, so I don't see returning false in the Try* API would help all that much. cc @steveharter 

> I think I might be able to use a `Pipe` and an `Utf8JsonReader` to do the job.
> Although it is quite a lot of code overhead, working with `SequencePosition` is awkward

Yep, that's true. Let me think about the scenario you presented and get back to you (I want to see how much effort is required if a motivated dev wants to support this case via the low-level reader). If you already have a pipe/utf8jsonreader-based implementation that works, please share.

**Comment by ericwj** [comment]
A quick and dirty implementation that works as long as the input is okay is [here](https://github.com/ericwj/DotnetCorefxIssue39843/blob/master/src/JsonArrayStream/JsonArrayDeserializer.cs).

> What would you expect TryDeserialize to do/return in your scenario where you have multiple json payloads within the file and it fails to deserialize?

I would like it to return `false` or some error status and not throw, because exceptions are wicked expensive. Something like this:

```CS
class JsonSerializer {
    public static bool TryDeserialize<T>(
        ref Utf8JsonReader reader,
        out T result,
        JsonSerializerOptions options);
}
```

The pipe reader code will search for `[`, `{` and their closing counterparts, where any unbalanced object tags will cause an exception calling `JsonSerializer.Deserialize<T>`. A naive way is just to try to deserialize for any `}` found, hoping it will match the first opening brace in the JSON remaining at that point, at the cost of a `try...catch` if the braces aren't balanced. Less naive might be to balance `{`'s with `}`'s, but that comes at the cost of more calls to `ReadOnlySequence<byte>.PositionOf(byte)` or yet higher code complexity and more awkward `SequencePosition` math and comparisons.

I don't immediately see the need for `SequencePosition` in favor of plain `int`s or `long`s at all, can the tracking of the `_object` field not be avoided by just having overloads that take some integral offset? Or it has to become something proper that is comparable and has suitable operators defined. But then still I don't see why I have to deal with them and drag that `_object` field along.

I seriously don't think using the `Utf8JsonReader` directly - that is, without calling into `JsonSerializer` - is beneficial except if performance is supercritical and you have to deserialize a lot of JSON. I don't think I might ever go that route except in a rare case when the schema is very, very limited and the volume to read/write justifies the effort.

**Comment by ahsonkhan** [comment]
Given this is currently by-design, and requires some thinking/design work to support this feature, moving to 5.0.
